### PR TITLE
feat: browser input

### LINKED ISSUE #15972: Browser input - Quo2 Component
Create the quo2 component browser input as per designs:

https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=4840-135087&t=rY9tKnH8a9ROUrMe-0

Should be implemented according to best practices
https://github.com/status-im/status-mobile/blob/develop/doc/new-guidelines.md

Include component spec tests
https://github.com/status-im/status-mobile/blob/develop/doc/component-tests-overview.md

Communicate with design team before starting to make sure there is no unknown requirements and to understand the interactions

Blur variant is out of scope

### PR DESCRIPTION
**Pull request by BalogunofAfrica**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

Reopens https://github.com/status-im/status-mobile/pull/16039
Fixes https://github.com/status-im/status-mobile/issues/15972

Implements the browser input component according to the [designs](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=4840-135087&t=1EwxozYZyGszVVME-0 "figma file").

The component is added to the quo preview in the profile settings.

 Preview 1

 Preview 2

 Preview 3

https://github.com/status-im/status-mobile/assets/45393944/53889c8a-bbf3-45ec-b1cf-0a24fab75381

Component overlayed with the design:

Screenshot 2023-07-04 at 22 59 12

status: ready

### CODE REVIEW

File: src/quo2/components/browser/browser_input/component_spec.cljs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            [test-helpers.component :as h]))
+
+(h/describe "Browser input"
+  (h/test "Renders empty in default state"
+    (h/render [browser-input/browser-input
+               {:on-change-text (h/mock-fn)
+                :value          ""}])
+    (h/is-truthy (h/get-by-label-text :browser-input)))
+
+  (h/test "Value updates on change"
```
Comment (J-Son89) [inline src/quo2/components/browser/browser_input/component_spec.cljs]: nit: test name is not accurate. It tests the on change was called not that the value was changed.

File: src/quo2/components/browser/browser_input/view.cljs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        set-default #(reset! state :default)
+        set-value   #(reset! value %)
+        ref         (atom nil)
+        clear-input (fn []
+                      (.clear ^js @ref)
+                      (reset! value ""))]
+    (fn [{:keys [disabled? blur? on-change-text customization-color
+                 on-clear on-focus on-blur theme get-ref use-ssl?
+                 favicon favicon-color favicon-size label]
+          :or   {customization-color :blue}
```
Comment (J-Son89) [inline src/quo2/components/browser/browser_input/view.cljs]: I think we don't need to set a default for `customization-color` it should always be provided - can you update in the preview screens if it is needed there 👍

File: src/quo2/components/browser/browser_input/view.cljs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+          [rn/text-input
+           (merge
+            clean-props
+            {:accessibility-label    :browser-input
+             :auto-capitalize        :none
+             :auto-correct           false
+             :cursor-color           (cursor-color customization-color theme)
+             :editable               (not disabled?)
+             :keyboard-appearance    (colors/theme-colors :light :dark theme)
+             :keyboard-type          "web-search"
```
Comment (J-Son89) [inline src/quo2/components/browser/browser_input/view.cljs]: can it `:web-search` instead?

File: src/quo2/core.cljs [hunk 4]
```diff
@@ -135,6 +136,9 @@
 (def predictive-keyboard quo2.components.buttons.predictive-keyboard.view/view)
 (def slide-button quo2.components.buttons.slide-button.view/view)
 
+;;;; BROWSER
```
Comment (J-Son89) [inline src/quo2/core.cljs:141]: 👍

File: src/status_im2/contexts/quo_preview/browser/browser_input.cljs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+       [quo/page-nav
+        {:align-mid?   true
+         :mid-section  {:type :text-only :main-text ""}
+         :left-section {:icon :i/arrow-left
+                        :on-press
+                        #(rf/dispatch [:navigate-back])}}]
+       [rn/flat-list
+        {:header                    [preview/customizer state descriptor]
+         :key-fn                    str
+         :keyboardShouldPersistTaps :always
```
Comment (J-Son89) [inline src/status_im2/contexts/quo_preview/browser/browser_input.cljs]: keyboardShouldPersistTaps -> keyboard-should-persist-taps

File: src/quo2/components/browser/browser_input/view.cljs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        set-active  #(reset! state :active)
+        set-default #(reset! state :default)
+        set-value   #(reset! value %)
+        ref         (atom nil)
+        clear-input (fn []
+                      (.clear ^js @ref)
+                      (reset! value ""))]
+    (fn [{:keys [disabled? blur? on-change-text customization-color
+                 on-clear on-focus on-blur theme get-ref use-ssl?
+                 favicon favicon-color favicon-size label]
```
Comment (J-Son89) [inline src/quo2/components/browser/browser_input/view.cljs]: the prop `label` is confusing me somewhat here. In your preview screen it seems to be a separate component that sits above the browser input. Is this correct, or should it be a different state of the browser input? e.g the domain name has resolved is now displaying in that format

For example here is a screen with it being used. 
Screenshot 2023-07-05 at 09 27 51
Perhaps you can verify this with the design team?

File: src/quo2/components/browser/browser_input/view.cljs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        value       (reagent/atom default-value)
+        set-active  #(reset! state :active)
+        set-default #(reset! state :default)
+        set-value   #(reset! value %)
+        ref         (atom nil)
+        clear-input (fn []
+                      (.clear ^js @ref)
+                      (reset! value ""))]
+    (fn [{:keys [disabled? blur? on-change-text customization-color
+                 on-clear on-focus on-blur theme get-ref use-ssl?
```
Comment (J-Son89) [inline src/quo2/components/browser/browser_input/view.cljs]: nit: - I think `locked?` might be an easier name then `use-ssl` for devs to find.

**Review by J-Son89 — APPROVED** [review]

File: src/quo2/components/browser/browser_input/style.cljs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+   :z-index            10})
+
+(defn text
+  []
+  (assoc (text/text-style {:size   :paragraph-1
+                           :weight :medium})
+         :color
+         (colors/theme-colors colors/neutral-100 colors/white)))
+
+(def rooot-container
```
Comment (ibrkhalil) [inline src/quo2/components/browser/browser_input/style.cljs]: Typo, should be root-container

File: src/quo2/components/browser/browser_input/view.cljs [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+(defn- clear-button
+  [{:keys [on-press blur? theme]}]
+  [rn/touchable-opacity
+   {:accessibility-label :browser-input-clear-button
+    :on-press            on-press
+    :style               style/clear-icon-container}
+   [icon/icon :i/clear
+    {:color (clear-icon-color blur? theme)
+     :size  20}]])
```
Comment (ibrkhalil) [inline src/quo2/components/browser/browser_input/view.cljs]: I think default size is 20 so this can be omitted, By no worries let's keep it.

**Review by ibrkhalil — COMMENTED** [review]
Should the browser default to a search engine like Google, Yahoo, Bing in case the user didn't add a specifc URL?
And should we allow him to alter the selection of the default engine?

File: src/quo2/components/browser/browser_input/view.cljs (hunk 9 above)
Comment (J-Son89) [reply src/quo2/components/browser/browser_input/view.cljs]: let's remove it if it's redundant 👍

**Comment by BalogunofAfrica** [comment]
> Should the browser default to a search engine like Google, Yahoo, Bing in case the user didn't add a specifc URL? And should we allow him to alter the selection of the default engine?

This PR only adds the input component for the browser and doesn't implement the actual search/query which is where I think the consideration would come in.

**Comment by cammellos** [comment]
@BalogunofAfrica linting fails:
```

Configuring Nix shell for target 'clojure'...

bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)

bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)

sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)

Node modules not created by Nix

Copying node_modules from Nix store:

 - /nix/store/asi99lnx45c0kdpmxcg5xmplwflnp12i-status-mobile-node-deps-1.23.0-patched

Done in: 2.052s

sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)

src/status_im2/contexts/quo_preview/browser/browser_input.cljs:24:22: warning: unused binding keyboard-show-listener

src/status_im2/contexts/quo_preview/browser/browser_input.cljs:27:22: warning: unused binding keyboard-hide-listener

linting took 8590ms, errors: 0, warnings: 2

make: *** [Makefile:306: lint] Error 2

script returned exit code 2
```

You might want to address the issues and run `make lint` afterward to make sure it's ok.
Thank you!

**Review by Francesca-G — APPROVED** [review]

**Comment by BalogunofAfrica** [comment]
@cammellos I have updated the PR with the fix

**Comment by cammellos** [comment]
@BalogunofAfrica thank you, lint passes!

Component tests are now failing:

```
PASS component-spec/quo2.components.drawers.permission_context.component_spec.js

FAIL component-spec/quo2.components.browser.browser_input.component_spec.js (197.484 s)

  ● Browser input › Input Label › Renders label text when specified



    expect(received).toBeTruthy()



    Received: null



      at test-helpers.component/query-by-label-text (test_helpers/component.cljs:189:27)

      at Object.<anonymous> (quo2/components/browser/browser_input/component_spec.cljs:28:21)



  ● Browser input › Input Label › Renders label favicon when specified alongside label text



    expect(received).toBeTruthy()



    Received: null



      at test-helpers.component/query-by-label-text (test_helpers/component.cljs:189:27)

      at Object.<anonymous> (quo2/components/browser/browser_input/component_spec.cljs:33:21)



  ● Browser input › Input Label › Renders lock icon when using ssl alongside label text



    expect(received).toBeTruthy()



    Received: null



      at test-helpers.component/query-by-label-text (test_helpers/component.cljs:189:27)

      at Object.<anonymous> (quo2/components/browser/browser_input/component_spec.cljs:38:21)

```

would you mind having a look please?

thank you

**Comment by BalogunofAfrica** [comment]
@cammellos Fixed test errors

**Comment by cammellos** [comment]
Thank you @BalogunofAfrica , amazing work!

File: src/quo2/components/browser/browser_input/style.cljs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+(def lock-icon-container
+  {:margin-left 2})
+
+(defn active-container
+  [display?]
+  {:align-items     :center
+   :flex-direction  :row
+   :justify-content :center
+   :opacity         (if display? 1 0)
+   :width           "100%"})
```
Comment (flexsurfer) [inline src/quo2/components/browser/browser_input/style.cljs]: width is redundant

File: src/quo2/components/browser/browser_input/style.cljs [hunk 11]
```diff
@@ -0,0 +1,57 @@
+(ns quo2.components.browser.browser-input.style
+  (:require [quo2.components.markdown.text :as text]
+            [quo2.foundations.colors :as colors]))
+
+(def clear-icon-container
+  {:align-items     :center
+   :height          20
+   :justify-content :center
+   :margin-left     8
+   :top             3.5
```
Comment (flexsurfer) [inline src/quo2/components/browser/browser_input/style.cljs]: should be integer

File: src/quo2/components/browser/browser_input/view.cljs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+             :style                  (style/input disabled?)})]
+          (when (seq @value)
+            [clear-button
+             {:blur?    blur?
+              :on-press (fn []
+                          (clear-input)
+                          (when on-clear (on-clear)))
+              :theme    theme}])]]))))
+
+
```
Comment (flexsurfer) [inline src/quo2/components/browser/browser_input/view.cljs]: extra line

File: src/quo2/components/browser/browser_input/view.cljs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                      (.clear ^js @ref)
+                      (reset! value ""))
+        focus-input (fn []
+                      (set-active)
+                      (.focus ^js @ref))]
+    (fn [{:keys [disabled? blur? on-change-text customization-color
+                 on-clear on-focus on-blur theme get-ref locked?
+                 favicon favicon-color favicon-size]
+          :as   props}]
+      (let [clean-props (apply dissoc props props-to-remove)]
```
Comment (flexsurfer) [inline src/quo2/components/browser/browser_input/view.cljs:89]: could you pls elaborate on why we need to remove props?

File: src/quo2/components/browser/browser_input/view.cljs [hunk 14]
```diff
@@ -0,0 +1,145 @@
+(ns quo2.components.browser.browser-input.view
+  (:require [quo2.components.icon :as icon]
+            [quo2.components.browser.browser-input.style :as style]
+            [react-native.core :as rn]
+            [reagent.core :as reagent]
+            [quo2.foundations.colors :as colors]
+            [quo2.theme :as quo2.theme]
```
Comment (flexsurfer) [inline src/quo2/components/browser/browser_input/view.cljs:7]: just `theme`, quo2 is redundant

File: src/quo2/components/browser/browser_input/view.cljs [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+    (fn [{:keys [disabled? blur? on-change-text customization-color
+                 on-clear on-focus on-blur theme get-ref locked?
+                 favicon favicon-color favicon-size]
+          :as   props}]
+      (let [clean-props (apply dissoc props props-to-remove)]
+        [rn/view {:style style/root-container}
+         (when (and (seq @value) (= @state :default))
+           [rn/touchable-opacity
+            {:style    style/default-container
+             :on-press (fn [] (focus-input))}
```
Comment (flexsurfer) [inline src/quo2/components/browser/browser_input/view.cljs]: fn [] is redundant just `focus-input`

**Review by flexsurfer — APPROVED** [review]

File: src/quo2/components/browser/browser_input/view.cljs (hunk 14 above)
Comment (J-Son89) [reply src/quo2/components/browser/browser_input/view.cljs:7]: actually it's not @flexsurfer. Here we are using `theme` prop as well so there is a likely chance these two references clash and there is unsafe uses. 
Similar to @ilmotta's pr: https://github.com/status-im/status-mobile/pull/16500

however we have been using `[quo2.theme :as quo.theme]` in other files

**Pull request merged**